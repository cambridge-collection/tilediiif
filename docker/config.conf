IMAGE_ORGANISATION=camdl
IMAGE_NAME=tilediiif-pipelines
IMAGE_VERSION=1.0.1
