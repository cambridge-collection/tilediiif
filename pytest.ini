[pytest]
addopts = --doctest-modules --ignore=tilediiif/server/wsgi_gunicorn_meinheld.py
