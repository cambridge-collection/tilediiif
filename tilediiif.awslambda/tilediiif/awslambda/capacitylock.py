import threading
from contextlib import contextmanager
from dataclasses import dataclass, field
from typing import Generator


@dataclass
class CapacityLock:
    """A lock that limits concurrent access to a resource, where each accessor
    uses variable-size fraction of the resource's capacity.

    For example, (and the use case in this module) CapacityLock can prevent too
    many concurrent threads consuming all available disk space. The CapacityLock
    is created with the available disk space as the total_capacity. Each thread
    calling acquire() specifies the amount of space it needs. Concurrent
    acquire() calls can proceed, so long as they wouldn't cause the
    total_capacity—and thus disk space—to be exceeded.
    """

    total_capacity: int
    _used_capacity: int = field(default=0)
    _condition: threading.Condition = field(default_factory=threading.Condition)

    def __post_init__(self) -> None:
        if self.total_capacity <= 0:
            raise ValueError("total_capacity must be > 0")

    @property
    def used_capacity(self) -> int:
        return self._used_capacity

    @contextmanager
    def acquire(self, capacity: int) -> Generator[None, None, None]:
        if capacity < 1:
            raise ValueError(f"capacity must be > 0 {capacity=}")
        with self._condition:
            # Always allow a single request to enter, even if it requests more
            # capacity than the limit, otherwise we'll deadlock. Our
            # responsibility is to prevent concurrent requests overflowing the
            # capacity, not to protect a resource from a single oversized
            # interaction — applications can check and handle this.
            while (
                self._used_capacity > 0
                and self._used_capacity + capacity > self.total_capacity
            ):
                self._condition.wait()
            self._used_capacity += capacity
            assert self._used_capacity > 0
        try:
            yield
        finally:
            with self._condition:
                self._used_capacity -= capacity
                assert self._used_capacity >= 0
                # Our cleared capacity could allow multiple acquire calls to
                # proceed.
                self._condition.notify_all()
