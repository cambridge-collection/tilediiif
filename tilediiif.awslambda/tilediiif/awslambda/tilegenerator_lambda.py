from __future__ import annotations

import json
import os
import tempfile
from concurrent.futures import Executor, ThreadPoolExecutor
from functools import partial
from pathlib import Path
from threading import BoundedSemaphore, Semaphore
from typing import TYPE_CHECKING, Optional, Sequence, TypedDict

import boto3
from boto3.s3.transfer import S3Transfer
from botocore.config import Config
from pydantic import BaseModel
from tilediiif.core.templates import Template, parse_template
from tilediiif.tools.dzi import parse_dzi_file
from tilediiif.tools.dzi_generation import ColourConfig, DZIConfig, JPEGConfig, save_dzi
from tilediiif.tools.infojson import get_id_url, info_json_from_dzi
from tilediiif.tools.tilelayout import (
    DEFAULT_FILE_PATH_TEMPLATE,
    create_dzi_tile_layout,
    create_file_via_hardlink,
    get_template_bindings,
    get_templated_dest_path,
    normalise_output_format,
)

if TYPE_CHECKING:
    import mypy_boto3_s3 as s3

MAX_CONCURRENT_SOURCES = 10


class SourceImageReference(BaseModel):
    bucket_name: str
    key: str
    identifier: str


class HandleDirectEvent(BaseModel):
    source_images: Sequence[SourceImageReference]
    iiif_base_url: str
    """The URL that IIIF images will be served under.

    e.g. a iiif_base_url of "https://iiif.example.com/images/" will result in
    an image with identifier "abc1" having URLs such as:

    - https://iiif.example.com/images/abc1/info.json
    - https://iiif.example.com/images/abc1/0,0,510,510/510,/0/default.jpeg
    """
    destination_bucket: str
    destination_key_prefix: Optional[str]


class SourceImageReferenceJson(TypedDict):
    bucket_name: str
    key: str
    identifier: str


class HandleDirectEventJson(TypedDict):
    source_images: Sequence[SourceImageReferenceJson]
    iiif_base_url: str
    destination_bucket: str
    destination_key_prefix: Optional[str]


class GeneratedTiles(TypedDict):
    identifier: str
    keys: Sequence[str]


def handle_direct(
    raw_event: HandleDirectEventJson, context: object
) -> Sequence[GeneratedTiles]:
    """Run the tile generator with domain-specific arguments.

    This handler is used to call the tilediiif tile generator when the caller
    has control over the arguments. For example, from a Step Function workflow,
    or manually from the AWS Console.
    """
    dzi_config = DZIConfig.from_environ()
    tile_encoding_config = JPEGConfig.from_environ()
    colour_config = ColourConfig.from_environ()

    raw_tile_path_template = os.environ.get(
        "TILE_PATH_TEMPLATE", DEFAULT_FILE_PATH_TEMPLATE
    )
    try:
        tile_path_template = parse_template(raw_tile_path_template)
    except ValueError as e:
        raise ValueError(f"invalid TILE_PATH_TEMPLATE: {e}") from e

    # TODO: validate this value. I think 20 should be OK for our use, as we use
    # s3transfer via download_file() upload_file(), and they use 10
    # threads/connections max each.
    boto_config = Config(max_pool_connections=30)
    s3client: s3.Client = boto3.client("s3", config=boto_config)
    s3_download = S3Transfer(client=s3client)
    s3_upload = S3Transfer(client=s3client)
    event = HandleDirectEvent.parse_obj(raw_event)
    source_executor = ThreadPoolExecutor()
    tile_executor = ThreadPoolExecutor()
    concurrent_source_limit = BoundedSemaphore(MAX_CONCURRENT_SOURCES)

    return [
        {"identifier": src.identifier, "keys": keys}
        for (src, keys) in source_executor.map(
            (
                lambda src: (
                    src,
                    fetch_generate_and_upload(
                        source_config=src,
                        id_base_url=event.iiif_base_url,
                        dzi_config=dzi_config,
                        tile_encoding_config=tile_encoding_config,
                        colour_config=colour_config,
                        tile_path_template=tile_path_template,
                        destination_bucket=event.destination_bucket,
                        destination_key_prefix=event.destination_key_prefix,
                        s3_download=s3_download,
                        s3_upload=s3_upload,
                        concurrent_source_limit=concurrent_source_limit,
                        tile_executor=tile_executor,
                    ),
                )
            ),
            event.source_images,
        )
    ]


def fetch_generate_and_upload(
    *,
    source_config: SourceImageReference,
    id_base_url: str,
    dzi_config: DZIConfig,
    tile_encoding_config: JPEGConfig,
    colour_config: ColourConfig,
    tile_path_template: Template,
    destination_bucket: str,
    destination_key_prefix: Optional[str],
    s3_download: S3Transfer,
    s3_upload: S3Transfer,
    concurrent_source_limit: Semaphore,
    tile_executor: Executor,
) -> Sequence[str]:
    with concurrent_source_limit, tempfile.TemporaryDirectory() as dir_string:
        working_dir = Path(dir_string)
        source_image = working_dir / "source_image"
        s3_download.download_file(
            bucket=source_config.bucket_name,
            key=source_config.key,
            filename=str(source_image),
        )
        iiif_tiles_dir = generate_tiles(
            source_config=source_config,
            source_image=source_image,
            id_base_url=id_base_url,
            dzi_config=dzi_config,
            tile_encoding_config=tile_encoding_config,
            colour_config=colour_config,
            tile_path_template=tile_path_template,
            working_dir=working_dir,
        )

        def upload(file: Path) -> str:
            relative_file = file.relative_to(iiif_tiles_dir)
            keyParts = [destination_key_prefix, source_config.identifier, relative_file]
            destination_key = str(Path(*[x for x in keyParts if x is not None]))
            s3_upload.upload_file(
                bucket=destination_bucket, key=destination_key, filename=str(file)
            )
            return destination_key

        return list(tile_executor.map(upload, sorted(iiif_tiles_dir.glob("**/*"))))


def generate_tiles(
    *,
    source_config: SourceImageReference,
    source_image: Path,
    id_base_url: str,
    dzi_config: DZIConfig,
    tile_encoding_config: JPEGConfig,
    colour_config: ColourConfig,
    tile_path_template: Template,
    working_dir: Path,
) -> Path:
    dest_dzi = working_dir / "image"
    save_dzi(
        src_image=source_image,
        dest_dzi=dest_dzi,
        dzi_config=dzi_config,
        tile_encoding_config=tile_encoding_config,
        colour_config=colour_config,
    )

    try:
        with open(f"{dest_dzi}.dzi", "rb") as f:
            dzi_meta = parse_dzi_file(f)
    except Exception as e:
        raise RuntimeError(f"Failed to parse generated .dzi file: {e}") from e

    get_dest_path = partial(
        get_templated_dest_path,
        tile_path_template,
        bindings_for_tile=partial(
            get_template_bindings, format=normalise_output_format(dzi_meta["format"])
        ),
    )

    iiif_tiles_dir = working_dir / "iiif-tiles"
    iiif_tiles_dir.mkdir()
    create_dzi_tile_layout(
        dzi_path=Path(f"{dest_dzi}.dzi"),
        dzi_meta=dzi_meta,
        get_dest_path=get_dest_path,
        create_file=create_file_via_hardlink,
        target_directory=iiif_tiles_dir,
    )

    id_url = get_id_url(id_base_url, source_config.identifier)
    json_meta = info_json_from_dzi(dzi_meta, id_url=id_url)
    info_json_file = iiif_tiles_dir / "info.json"
    info_json_file.write_text(json.dumps(json_meta))

    return iiif_tiles_dir
