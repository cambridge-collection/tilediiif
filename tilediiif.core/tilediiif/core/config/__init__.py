from .core import (
    BaseConfig,
    Config,
    ConfigProperty,
    EnvironmentConfigMixin,
    JSONConfigMixin,
    ParseResult,
    TOMLConfigMixin,
)
from .exceptions import ConfigError, ConfigParseError

__all__ = [
    "BaseConfig",
    "Config",
    "ConfigProperty",
    "ConfigError",
    "ConfigParseError",
    "EnvironmentConfigMixin",
    "JSONConfigMixin",
    "ParseResult",
    "TOMLConfigMixin",
]
