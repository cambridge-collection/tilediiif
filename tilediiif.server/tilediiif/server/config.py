import json
from enum import Enum
from pathlib import Path

from tilediiif.core.config import Config, ConfigProperty
from tilediiif.core.config.properties import EnumConfigProperty

with open(Path(__file__).parent / "config-schema.json") as f:
    CONFIG_SCHEMA = json.load(f)


class FileTransmissionType(Enum):
    DIRECT = "direct"
    INDIRECT = "indirect"


class ConfigValueEnvars(Enum):
    DATA_PATH = "TILEDIIIF_SERVER_DATA_PATH"
    FILE_TRANSMISSION = "TILEDIIIF_SERVER_FILE_TRANSMISSION"
    SENDFILE_HEADER_NAME = "TILEDIIIF_SERVER_SENDFILE_HEADER_NAME"
    IMAGE_PATH_TEMPLATE = "TILEDIIIF_SERVER_IMAGE_PATH_TEMPLATE"
    INFO_JSON_PATH_TEMPLATE = "TILEDIIIF_SERVER_INFO_JSON_PATH_TEMPLATE"


ConfigValueEnvars.envar_names = frozenset(e.name for e in ConfigValueEnvars)


class ServerConfig(Config):
    json_schema = CONFIG_SCHEMA

    property_definitions = [
        ConfigProperty(
            name="image_path_template",
            default="{identifier}/{region}-{size}-{rotation}-{quality}.{format}",
            envar_name=ConfigValueEnvars.IMAGE_PATH_TEMPLATE.value,
            json_path="tilediiif.server.image-path-template",
        ),
        ConfigProperty(
            name="info_json_path_template",
            default="{identifier}/info.json",
            envar_name=ConfigValueEnvars.INFO_JSON_PATH_TEMPLATE.value,
            json_path="tilediiif.server.info-json-path-template",
        ),
        ConfigProperty(
            name="sendfile_header_name",
            default="X-Accel-Redirect",
            envar_name=ConfigValueEnvars.SENDFILE_HEADER_NAME.value,
            json_path=(
                "tilediiif.server.indirect-file-transmission.sendfile-header-name"
            ),
        ),
        ConfigProperty(
            name="data_path",
            default=".",
            envar_name=ConfigValueEnvars.DATA_PATH.value,
            json_path="tilediiif.server.data-path",
        ),
        EnumConfigProperty(
            name="file_transmission",
            enum_cls=FileTransmissionType,
            default=FileTransmissionType.DIRECT,
            envar_name=ConfigValueEnvars.FILE_TRANSMISSION.value,
            json_path="tilediiif.server.file-transmission",
        ),
    ]
