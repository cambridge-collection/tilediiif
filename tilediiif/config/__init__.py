from .core import (
    BaseConfig,
    Config,
    ConfigError,
    ConfigParseError,
    ConfigProperty,
    EnvironmentConfigMixin,
    JSONConfigMixin,
    ParseResult,
    TOMLConfigMixin,
)

__all__ = [
    "BaseConfig",
    "Config",
    "ConfigError",
    "ConfigParseError",
    "ConfigProperty",
    "EnvironmentConfigMixin",
    "JSONConfigMixin",
    "ParseResult",
    "TOMLConfigMixin",
]
