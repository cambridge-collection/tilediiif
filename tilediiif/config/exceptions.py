from typing import TYPE_CHECKING

from attr import dataclass

if TYPE_CHECKING:
    from tilediiif.config import BaseConfig, ConfigProperty


class ConfigError(ValueError):
    pass


class ConfigParseError(ConfigError):
    pass


class ConfigValidationError(ConfigError):
    pass


@dataclass
class ConfigValueNotPresent(LookupError):
    config: "BaseConfig"
    property: "ConfigProperty"

    def __str__(self):
        return (
            f"No value for config property {self.property.name!r} exists in config: "
            f"{self.config}"
        )


class CLIValueNotFound(LookupError):
    pass


class InvalidCLIUsageConfigError(ConfigError):
    pass
