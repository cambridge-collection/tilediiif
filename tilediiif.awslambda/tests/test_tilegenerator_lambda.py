from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING, Callable, Sequence

import boto3
import pytest
from moto import mock_s3  # type: ignore
from syrupy.assertion import SnapshotAssertion
from tilediiif.core.config.exceptions import ConfigParseError

ROOT = Path(__file__).parents[2]
TEST_IMAGES = ROOT / "tilediiif.tools/integration_tests/data/images"

if TYPE_CHECKING:
    from mypy_boto3_s3 import Client

    from tilediiif.awslambda.tilegenerator_lambda import (
        GeneratedTiles,
        HandleDirectEventJson,
        SourceImageReferenceJson,
    )


@pytest.fixture
def aws_credentials(monkeypatch: pytest.MonkeyPatch):
    """Mocked AWS Credentials for moto."""
    monkeypatch.setenv("AWS_ACCESS_KEY_ID", "testing")
    monkeypatch.setenv("AWS_SECRET_ACCESS_KEY", "testing")
    monkeypatch.setenv("AWS_SECURITY_TOKEN", "testing")
    monkeypatch.setenv("AWS_SESSION_TOKEN", "testing")
    monkeypatch.setenv("AWS_DEFAULT_REGION", "us-east-1")


@pytest.fixture
def mock_s3_enabled(aws_credentials):
    # mock_s3() mocks the boto3 s3 APIs so that they can be used without without
    # triggering network requests. e.g. calls to store and fetch data work as
    # normal, but only with changes made in the tests being visible.
    with mock_s3():
        yield


@pytest.fixture
def handle_direct(mock_s3_enabled):
    from tilediiif.awslambda.tilegenerator_lambda import handle_direct

    return handle_direct


@pytest.fixture
def s3client(mock_s3_enabled) -> Client:
    return boto3.client("s3")


@pytest.fixture
def source_bucket(s3client: Client) -> str:
    name = "SourceBucket"
    s3client.create_bucket(Bucket=name)
    return name


@pytest.fixture
def dest_bucket(s3client: Client) -> str:
    name = "DestBucket"
    s3client.create_bucket(Bucket=name)
    return name


@pytest.fixture
def source_pears_small(s3client: Client, source_bucket: str) -> dict[str, str]:
    s3client.upload_file(
        Bucket=source_bucket,
        Filename=str(TEST_IMAGES / "pears_small_srgb.jpg"),
        Key="A123-Thing/PHOTO-42.jpg",
    )
    return {
        "bucket_name": source_bucket,
        "key": "A123-Thing/PHOTO-42.jpg",
        "identifier": "A123-Thing_42",
    }


@pytest.fixture
def source_sunset_p3(s3client: Client, source_bucket: str) -> SourceImageReferenceJson:
    s3client.upload_file(
        Bucket=source_bucket,
        Filename=str(TEST_IMAGES / "Sunset-P3.jpg"),
        Key="A123-Thing/PHOTO-55.jpg",
    )
    return {
        "bucket_name": source_bucket,
        "key": "A123-Thing/PHOTO-55.jpg",
        "identifier": "A123-Thing_55",
    }


@pytest.fixture
def input_event(
    dest_bucket: str,
    source_pears_small: SourceImageReferenceJson,
    source_sunset_p3: SourceImageReferenceJson,
) -> HandleDirectEventJson:
    return {
        "source_images": [source_pears_small, source_sunset_p3],
        "iiif_base_url": "https://iiif.testing.test/images/",
        "destination_bucket": "DestBucket",
        "destination_key_prefix": "ExcellentImages",
    }


def test_handle_direct__default_options(
    handle_direct: Callable[[HandleDirectEventJson, object], Sequence[GeneratedTiles]],
    input_event: HandleDirectEventJson,
    snapshot: SnapshotAssertion,
) -> None:
    results = handle_direct(input_event, {})

    assert [result["identifier"] for result in results] == [
        "A123-Thing_42",
        "A123-Thing_55",
    ]
    assert results == snapshot


def test_handle_direct__non_default_options(
    handle_direct: Callable[[HandleDirectEventJson, object], Sequence[GeneratedTiles]],
    input_event: HandleDirectEventJson,
    monkeypatch: pytest.MonkeyPatch,
    snapshot: SnapshotAssertion,
) -> None:
    monkeypatch.setenv("DZI_TILES_DZI_TILE_SIZE", "510")  # 512 with overlap

    input_event["destination_key_prefix"] = "MyPrefix/Stuff"
    results = handle_direct(input_event, {})

    assert [result["identifier"] for result in results] == [
        "A123-Thing_42",
        "A123-Thing_55",
    ]
    assert results == snapshot


@pytest.mark.parametrize(
    "envar_name, envar_value, err_message",
    [
        (
            "DZI_TILES_JPEG_QUALITY",
            "invalid",
            (
                "Failed to parse config value for quality: Failed to parse config value"
                " for quality: invalid literal for int() with base 10: 'invalid"
            ),
        ),
        (
            "DZI_TILES_COLOUR_TRANSFORM_INTENT",
            "invalid",
            "'invalid' is not a valid RenderingIntent",
        ),
        (
            "DZI_TILES_DZI_TILE_SIZE",
            "invalid",
            (
                "Failed to parse config value for tile_size: invalid literal for int()"
                " with base 10: 'invalid'"
            ),
        ),
    ],
)
def test_handle_direct__reads_configuration_from_envars(
    envar_name: str,
    envar_value: str,
    err_message: str,
    handle_direct: Callable[[HandleDirectEventJson, object], Sequence[GeneratedTiles]],
    input_event: HandleDirectEventJson,
    monkeypatch: pytest.MonkeyPatch,
) -> None:
    monkeypatch.setenv(envar_name, envar_value)

    with pytest.raises(ConfigParseError) as exc_info:
        handle_direct(input_event, {})

    assert err_message in str(exc_info.value)
