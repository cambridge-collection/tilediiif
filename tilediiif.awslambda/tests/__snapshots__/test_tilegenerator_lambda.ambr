# name: test_handle_direct__default_options
  list([
    dict({
      'identifier': 'A123-Thing_42',
      'keys': list([
        'ExcellentImages/A123-Thing_42/0,0,1000,750-250,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/0,0,254,254-254,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/0,0,508,508-254,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/0,254,254,254-254,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/0,508,254,242-254,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/0,508,508,242-254,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/254,0,254,254-254,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/254,254,254,254-254,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/254,508,254,242-254,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/508,0,254,254-254,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/508,0,492,508-246,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/508,254,254,254-254,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/508,508,254,242-254,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/508,508,492,242-246,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/762,0,238,254-238,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/762,254,238,254-238,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/762,508,238,242-238,-0-default.jpg',
        'ExcellentImages/A123-Thing_42/info.json',
      ]),
    }),
    dict({
      'identifier': 'A123-Thing_55',
      'keys': list([
        'ExcellentImages/A123-Thing_55/0,0,254,254-254,-0-default.jpg',
        'ExcellentImages/A123-Thing_55/0,0,397,508-199,-0-default.jpg',
        'ExcellentImages/A123-Thing_55/0,0,397,600-100,-0-default.jpg',
        'ExcellentImages/A123-Thing_55/0,254,254,254-254,-0-default.jpg',
        'ExcellentImages/A123-Thing_55/0,508,254,92-254,-0-default.jpg',
        'ExcellentImages/A123-Thing_55/0,508,397,92-199,-0-default.jpg',
        'ExcellentImages/A123-Thing_55/254,0,143,254-143,-0-default.jpg',
        'ExcellentImages/A123-Thing_55/254,254,143,254-143,-0-default.jpg',
        'ExcellentImages/A123-Thing_55/254,508,143,92-143,-0-default.jpg',
        'ExcellentImages/A123-Thing_55/info.json',
      ]),
    }),
  ])
# ---
# name: test_handle_direct__non_default_options
  list([
    dict({
      'identifier': 'A123-Thing_42',
      'keys': list([
        'MyPrefix/Stuff/A123-Thing_42/0,0,1000,750-500,-0-default.jpg',
        'MyPrefix/Stuff/A123-Thing_42/0,0,510,510-510,-0-default.jpg',
        'MyPrefix/Stuff/A123-Thing_42/0,510,510,240-510,-0-default.jpg',
        'MyPrefix/Stuff/A123-Thing_42/510,0,490,510-490,-0-default.jpg',
        'MyPrefix/Stuff/A123-Thing_42/510,510,490,240-490,-0-default.jpg',
        'MyPrefix/Stuff/A123-Thing_42/info.json',
      ]),
    }),
    dict({
      'identifier': 'A123-Thing_55',
      'keys': list([
        'MyPrefix/Stuff/A123-Thing_55/0,0,397,510-397,-0-default.jpg',
        'MyPrefix/Stuff/A123-Thing_55/0,0,397,600-199,-0-default.jpg',
        'MyPrefix/Stuff/A123-Thing_55/0,510,397,90-397,-0-default.jpg',
        'MyPrefix/Stuff/A123-Thing_55/info.json',
      ]),
    }),
  ])
# ---
